### PR TITLE
Fix continues performance

### PR DESCRIPTION
**Pull request by dkropachev**
Introduce continues_hash that is used to find continues events

https://trello.com/c/R0BXSqmQ/4270-dblogreader-improve-continuous-events-performance
## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

File: sdcm/sct_events/continuous_event.py [hunk 1]
```diff
@@ -44,87 +46,59 @@ def add_event(self, event: ContinuousEvent):
...
             raise ContinuousEventRegistryException(msg)
 
-        if self._find_event_by_id(event.event_id):
+        if self.find_events_by_id(event.event_id):
             msg = f"Event with id: {event.event_id} is already present. Event ids in the registry must be unique."
             raise ContinuousEventRegistryException(msg)
 
         self.continuous_events.append(event)
 
+        if present_event := self.hashed_continuous_events.get(event.continues_hash):
```
Comment (bentsi) [inline sdcm/sct_events/continuous_event.py:55]: `continues_hash` -> `continuous_hash`

File: sdcm/sct_events/continuous_event.py [hunk 2]
```diff
@@ -134,6 +108,27 @@ def __init__(self,
...
+        if "shard" in data:
+            data["shard"] = int(data["shard"])
+        return str(uuid.uuid3(
+            cls.continues_hash_namespace,
+            str({attr_value: data.get(attr_value, None) for attr_value in cls.continues_hash_fields})
+        ))
+
+    @cached_property
+    def continues_hash_dict(self) -> dict[str, str]:
+        return {attr_value: getattr(self, attr_value, None) for attr_value in self.continues_hash_fields}
```
Comment (bentsi) [inline sdcm/sct_events/continuous_event.py:130]: `attr_value` -> `attr_name`

File: sdcm/sct_events/continuous_event.py [hunk 3]
```diff
@@ -134,6 +108,27 @@ def __init__(self,
...
+            str(self.continues_hash_dict)
+        ))
+
+    @classmethod
+    def get_continues_hash_from_dict(cls, data: dict):
+        if "shard" in data:
+            data["shard"] = int(data["shard"])
+        return str(uuid.uuid3(
+            cls.continues_hash_namespace,
+            str({attr_value: data.get(attr_value, None) for attr_value in cls.continues_hash_fields})
```
Comment (bentsi) [inline sdcm/sct_events/continuous_event.py:125]: `attr_value` -> `attr_name`

**Review by bentsi — CHANGES_REQUESTED** [review]
add default behavior when continuous hash fields were not set in a new event

**Comment by dkropachev** [comment]
Provision failed due to the - https://github.com/scylladb/scylla-cluster-tests/issues/4252, has nothing to do with this PR

File: sdcm/sct_events/continuous_event.py [hunk 4]
```diff
@@ -33,98 +34,66 @@ class ContinuousEventRegistryException(BaseException):
...
 
-        if self._find_event_by_id(event.event_id):
+        if self.find_events_by_id(event.event_id):
             msg = f"Event with id: {event.event_id} is already present. Event ids in the registry must be unique."
             raise ContinuousEventRegistryException(msg)
 
-        self.continuous_events.append(event)
+        if present_event := self.hashed_continuous_events.get(event.continuous_hash):
+            LOGGER.debug(
+                'Continues event %s with hash %s (%s) is getting replaces with another event %s',
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/sct_events/continuous_event.py:54]: fixed

File: sdcm/sct_events/continuous_event.py [hunk 5]
```diff
@@ -33,98 +34,66 @@ class ContinuousEventRegistryException(BaseException):
...
+    def continuous_events(self) -> Iterable[ContinuousEvent]:
+        return self.hashed_continuous_events.values()
 
     def add_event(self, event: ContinuousEvent):
         if not issubclass(type(event), ContinuousEvent):
             msg = f"Event: {event} is not a ContinuousEvent"
             raise ContinuousEventRegistryException(msg)
 
-        if self._find_event_by_id(event.event_id):
+        if self.find_events_by_id(event.event_id):
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/sct_events/continuous_event.py]: removed

File: sdcm/sct_events/continuous_event.py [hunk 6]
```diff
@@ -33,98 +34,66 @@ class ContinuousEventRegistryException(BaseException):
...
-        registry_filter = ContinuousRegistryFilter(registry=self.continuous_events)
-
-        return registry_filter
-
-    def _find_event_by_id(self, event_id: Union[uuid.UUID, str]) -> List[ContinuousEvent]:
-        event_filter = self.get_registry_filter()
-        found_events = event_filter.filter_by_id(event_id).get_filtered()
-
-        return found_events
+    def find_events_by_id(self, event_id: Union[uuid.UUID, str]) -> List[ContinuousEvent]:
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/sct_events/continuous_event.py]: removed

File: unit_tests/test_sct_events_continuous_events_registry.py [hunk 7]
```diff
@@ -63,48 +61,14 @@ def test_adding_a_non_continuous_event_raises_error(self,
             registry.add_event(info_event)
 
     def test_get_event_by_id(self,
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply unit_tests/test_sct_events_continuous_events_registry.py]: removed

File: sdcm/sct_events/continuous_event.py [hunk 8]
```diff
@@ -33,98 +34,66 @@ class ContinuousEventRegistryException(BaseException):
...
+        super().__init_subclass__(*args, **kwargs)
 
     def __init__(self,
                  severity: Severity = Severity.UNKNOWN,
-                 publish_event: bool = True):
+                 publish_event: bool = True, errors: list[str] = None):
         super().__init__(severity=severity)
-        self.log_file_name = None
-        self.errors = []
+        if not errors:
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/sct_events/continuous_event.py]: done

**Review by KnifeyMoloko — DISMISSED** [review]
Looks good :)

**Review by bentsi — CHANGES_REQUESTED** [review]
- let's simplify the logic and use tuples instead of generating hashes
- let's remove register/unregister methods and call add_event and del_event directly on the event registry object in `begin_event()` and `end_event()` methods

**Review by bentsi — APPROVED** [review]

**Pull request merged**